sql_connectors/storage.py

```diff
@@ -0,0 +1,209 @@
+                "default_env",
+                "default_schema",
+                "default_reflect",
+            ]
+            return [key for key in keys if key not in non_envs]
+
+        return get_available_envs
+
+    def _get_username(self, conf):
+        """Get username and prompt user if necessary. Set to None if it's empty."""
+        username = conf.get("username", None)
+        allowed_hosts = conf.get("allowed_hosts", [])
+        if username is None or (
+            len(allowed_hosts) > 0 and not os.uname()[1] in allowed_hosts
+        ):
+            username = input("Username: ")
+        if username == "":
+            username = None
+        return username
+
+    def _get_password(self, conf):
+        """Get password and prompt user if necessary. Set to None if it's empty."""
+        password = conf.get("password", None)
+        allowed_hosts = conf.get("allowed_hosts", [])
+        if password is None or (
+            len(allowed_hosts) > 0 and not os.uname()[1] in allowed_hosts
+        ):
+            password = getpass("Password: ")
+        if password == "":
+            password = None
+        return password
+
+    def get_config_defaults(self, conf):
+        """Return the default_env, default_schema, and default_reflect from a
+        config file.
+
+        :param str path: Path for config file
+        """
+        defaults = {
+            "default_env": "default",
+            "default_schema": None,
+            "default_reflect": False,
+        }
+        return dict((k, conf.get(k, defaults[k])) for k in defaults)
+
+    def get_client_factory(
+        self, conf, default_env="default", default_schema=None, default_reflect=False
+    ):
+        """Wrapper function to create a :any:`get_client` function using the given
+        ``config`` and setting the given defaults. This should be used in the submodule
+        for each data source.
+
+        :param str conf: Name of config file without the extension
+        :param str default_env: Set default environment to use (Default value = 'default')
+        :param str default_schema: Set default schema to use  (Default value = None)
+        :param bool default_reflect: Set default for reflect  (Default value = False)
+        """
+
+        @extend_docs(SqlClient.__init__)
+        def get_client(
+            env=default_env,
+            default_schema=default_schema,
+            reflect=default_reflect,
+            **kwargs
+        ):
+            """Get a :any:`SqlClient` for the specified
+            environment. Defaults are based on what was passed to :any:`get_client_factory`.
+
+            See :any:`SqlClient.__init__` for params:
+            """
+            return SqlClient(
+                self.parse_config(conf, env), default_schema, reflect, **kwargs
+            )
+
+        return memoized(get_client, signature_preserving=True)
+
+
+class LocalStorage(Storage):
+    def full_path(self, sub_path):
```

Comment:
Maybe make this a @staticmethod with default `sub_path=""` for easy troubleshooting. I actually didn't see this method before my previous comment, and I think it makes sense to have this here vs. in `config_util` since all of the local logic is now consolidated. Just thinking of ways to make it easier for the end user to troubleshoot stuff with their path

Reply:
not sure this could be a `staticmethod` since it references `self.path_or_uri`. The alternative is to have the function take in `base_path` and `sub_path`. That being said, wouldn't they want to instantiate the class first before trying to de-bug?

Or, would having a config module with a "default config" traitlet help with this?